README.md
# avtr-io

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

API Wrapper for avtr.io

## Install

Via Composer

``` bash
$ composer require jamosaur/avtr-io
```

## Usage

``` php
require Jamosaur\Avtr;
$avtr = new Avtr('James Wallen-Jones');

echo '<img src="' . $avtr->toUrl() . '">';
```

## Available Options
```php
// Construct with email.
$avtr = new Avtr('j.wallen.jones@googlemail.com');

// Construct with name.
$avtr = new Avtr('James Wallen-Jones');

// Construct with initials.
$avtr = new Avtr('JW');

// Image Format. (png, jpg, gif)
$avtr->format(string);

// Setting a first name.
$avtr->firstName(string);

// Setting a last name.
$avtr->lastName(string);

// Setting the letter count. (1, 2)
// Anything below 1 will default to 1.
// Anything above 2 will default to 2.
$avtr->letterCount(int);

// Setting the background colour.
// Values below 0 will default to 0.
// Values above 255 will default to 255.
// Alpha value is 0.0 - 1.0
$avtr->background(r, g, b, a);

// Setting the image size.
// Values below 0 will default to 100
$avtr->size(500);

// Setting rounded corners.
$avtr->roundedCorners(bool);

// Setting Shape. (square, circle)
$avtr->shape(string);

// Setting theme. (material, flat)
$avtr->theme(string);

// Setting text Case. (lower, upper, title)
$avtr->textCase(string);

// Setting text colour.
// Values below 0 will default to 0.
// Values above 255 will default to 255.
// Alpha value is 0.0 - 1.0
$avtr->color(r, g, b, a);

// Setting font weight. (100-900)
// Values below 100 default to 100
// Values above 900 default to 900
$avtr->fontWeight(100);

// Setting font. (open-sans, source-sans-pro, roboto)
$avtr->font(string);

// Return a URL
$avtr->toUrl();

```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CONDUCT](CONDUCT.md) for details.

## Security

If you discover any security related issues, please email j.wallen.jones@googlemail.com instead of using the issue tracker.

## Credits

- [James Wallen-Jones][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/Jamosaur/avtr-io.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/jamosaur/avtr-io/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/Jamosaur/avtr-io.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/jamosaur/avtr-io.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/Jamosaur/avtr-io.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/Jamosaur/avtr-io
[link-travis]: https://travis-ci.org/jamosaur/avtr-io
[link-scrutinizer]: https://scrutinizer-ci.com/g/Jamosaur/avtr-io/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/Jamosaur/avtr-io
[link-downloads]: https://packagist.org/packages/Jamosaur/avtr-io
[link-author]: https://github.com/jamosaur
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All Notable changes to `avtr-io` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## 1.0 - 2017-03-31

### Added
- Initial Release

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


tests/AvtrTest.php
<?php

namespace Jamosaur\Avtr;

use Jamosaur\Avtr\Exceptions\InvalidFontException;
use Jamosaur\Avtr\Exceptions\InvalidFormatException;
use Jamosaur\Avtr\Exceptions\InvalidShapeException;
use Jamosaur\Avtr\Exceptions\InvalidTextCaseException;
use Jamosaur\Avtr\Exceptions\InvalidThemeException;

class AvtrTest extends \PHPUnit_Framework_TestCase
{
    public function testConstructingWithEmail()
    {
        $avtr = new Avtr('email@test.com');
        $this->assertEquals(Avtr::URL . 'png?email=email%40test.com', $avtr->toUrl());
    }

    public function testConstructingWithInitials()
    {
        $avtr = new Avtr('JW');
        $this->assertEquals(Avtr::URL . 'png?initials=JW', $avtr->toUrl());
    }

    public function testConstructingWithFullName()
    {
        $avtr = new Avtr('James Wallen-Jones');
        $this->assertEquals(Avtr::URL . 'png?first_name=James&last_name=Wallen-Jones', $avtr->toUrl());
    }

    public function testConstructingWithFirstName()
    {
        $avtr = new Avtr('James');
        $this->assertEquals(Avtr::URL . 'png?initials=James', $avtr->toUrl());
    }

    public function testConstructingWithNothing()
    {
        $avtr = new Avtr;
        $this->assertEquals(Avtr::URL . 'png', $avtr->toUrl());
    }

    public function testSettingFormat()
    {
        $avtr = new Avtr;
        $this->assertEquals(Avtr::URL . 'png', $avtr->toUrl());
        $avtr->format('jpg');
        $this->assertEquals(Avtr::URL . 'jpg', $avtr->toUrl());
        $avtr->format('gif');
        $this->assertEquals(Avtr::URL . 'gif', $avtr->toUrl());
    }

    public function testSettingInvalidFormatThrowsException()
    {
        $this->expectException(InvalidFormatException::class);
        (new Avtr)->format('webm');
    }

    public function testSettingFirstName()
    {
        $avtr = (new Avtr)->firstName('James');
        $this->assertEquals(Avtr::URL . 'png?first_name=James', $avtr->toUrl());
    }

    public function testSettingLastName()
    {
        $avtr = (new Avtr)->lastName('Wallen-Jones');
        $this->assertEquals(Avtr::URL . 'png?last_name=Wallen-Jones', $avtr->toUrl());
    }

    public function testSettingLetterCount()
    {
        $avtr = (new Avtr)->letterCount(1);
        $this->assertEquals(Avtr::URL . 'png?letter_count=1', $avtr->toUrl());
        $avtr->letterCount(2);
        $this->assertEquals(Avtr::URL . 'png?letter_count=2', $avtr->toUrl());
        $avtr->letterCount(420);
        $this->assertEquals(Avtr::URL . 'png?letter_count=2', $avtr->toUrl());
        $avtr->letterCount(-69);
        $this->assertEquals(Avtr::URL . 'png?letter_count=2', $avtr->toUrl());
    }

    public function testSettingBackground()
    {
        $avtr = (new Avtr)->background(100, 100, 100, 0.5);
        $this->assertEquals(Avtr::URL . 'png?background=rgba%28100%2C100%2C100%2C0.5%29', $avtr->toUrl());
        $avtr->background(255, 255, 255);
        $this->assertEquals(Avtr::URL . 'png?background=rgba%28255%2C255%2C255%2C1%29', $avtr->toUrl());
        $avtr->background(1000, 1987234, 23846782364);
        $this->assertEquals(Avtr::URL . 'png?background=rgba%28255%2C255%2C255%2C1%29', $avtr->toUrl());
        $avtr->background(-1000, -1987234, -23846782364);
        $this->assertEquals(Avtr::URL . 'png?background=rgba%280%2C0%2C0%2C1%29', $avtr->toUrl());
        $avtr->background(-1000, -1987234, -23846782364, -12);
        $this->assertEquals(Avtr::URL . 'png?background=rgba%280%2C0%2C0%2C0%29', $avtr->toUrl());
        $avtr->background(-1000, -1987234, -23846782364, 12);
        $this->assertEquals(Avtr::URL . 'png?background=rgba%280%2C0%2C0%2C1%29', $avtr->toUrl());
        $avtr->background(-1000, -1987234, -23846782364, 0.5);
        $this->assertEquals(Avtr::URL . 'png?background=rgba%280%2C0%2C0%2C0.5%29', $avtr->toUrl());
    }

    public function testSettingSize()
    {
        $avtr = (new Avtr)->size(500);
        $this->assertEquals(Avtr::URL . 'png?size=500', $avtr->toUrl());
        $avtr->size(-100);
        $this->assertEquals(Avtr::URL . 'png?size=100', $avtr->toUrl());
    }

    public function testSettingRoundedCorners()
    {
        $avtr = (new Avtr)->roundedCorners(true);
        $this->assertEquals(Avtr::URL . 'png?rounded_corners=1', $avtr->toUrl());
        $avtr->roundedCorners(false);
        $this->assertEquals(Avtr::URL . 'png', $avtr->toUrl());
    }

    public function testSettingShapes()
    {
        $avtr = (new Avtr)->shape('square');
        $this->assertEquals(Avtr::URL . 'png?shape=square', $avtr->toUrl());
        $avtr->shape('circle');
        $this->assertEquals(Avtr::URL . 'png?shape=circle', $avtr->toUrl());
        $this->expectException(InvalidShapeException::class);
        $avtr->shape('triangle');
    }

    public function testSettingTheme()
    {
        $avtr = (new Avtr)->theme('material');
        $this->assertEquals(Avtr::URL . 'png?theme=material', $avtr->toUrl());
        $avtr->theme('flat');
        $this->assertEquals(Avtr::URL . 'png?theme=flat', $avtr->toUrl());
        $this->expectException(InvalidThemeException::class);
        $avtr->theme('metro');
    }

    public function testSettingFontCase()
    {
        $avtr = (new Avtr)->textCase('lower');
        $this->assertEquals(Avtr::URL . 'png?text_case=lower', $avtr->toUrl());
        $avtr->textCase('upper');
        $this->assertEquals(Avtr::URL . 'png?text_case=upper', $avtr->toUrl());
        $avtr->textCase('title');
        $this->assertEquals(Avtr::URL . 'png?text_case=title', $avtr->toUrl());
        $this->expectException(InvalidTextCaseException::class);
        $avtr->textCase('camel');
    }

    public function testSettingTextColor()
    {
        $avtr = (new Avtr)->color(100, 100, 100, 0.5);
        $this->assertEquals(Avtr::URL . 'png?text_color=rgba%28100%2C100%2C100%2C0.5%29', $avtr->toUrl());
        $avtr->color(255, 255, 255);
        $this->assertEquals(Avtr::URL . 'png?text_color=rgba%28255%2C255%2C255%2C1%29', $avtr->toUrl());
        $avtr->color(1000, 1987234, 23846782364);
        $this->assertEquals(Avtr::URL . 'png?text_color=rgba%28255%2C255%2C255%2C1%29', $avtr->toUrl());
        $avtr->color(-1000, -1987234, -23846782364);
        $this->assertEquals(Avtr::URL . 'png?text_color=rgba%280%2C0%2C0%2C1%29', $avtr->toUrl());
        $avtr->color(-1000, -1987234, -23846782364, -12);
        $this->assertEquals(Avtr::URL . 'png?text_color=rgba%280%2C0%2C0%2C0%29', $avtr->toUrl());
        $avtr->color(-1000, -1987234, -23846782364, 12);
        $this->assertEquals(Avtr::URL . 'png?text_color=rgba%280%2C0%2C0%2C1%29', $avtr->toUrl());
        $avtr->color(-1000, -1987234, -23846782364, 0.5);
        $this->assertEquals(Avtr::URL . 'png?text_color=rgba%280%2C0%2C0%2C0.5%29', $avtr->toUrl());
    }

    public function testSettingFontWeight()
    {
        $avtr = (new Avtr)->fontWeight(100);
        $this->assertEquals(Avtr::URL . 'png?font_weight=100', $avtr->toUrl());
        $avtr->fontWeight(200);
        $this->assertEquals(Avtr::URL . 'png?font_weight=200', $avtr->toUrl());
        $avtr->fontWeight(300);
        $this->assertEquals(Avtr::URL . 'png?font_weight=300', $avtr->toUrl());
        $avtr->fontWeight(400);
        $this->assertEquals(Avtr::URL . 'png?font_weight=400', $avtr->toUrl());
        $avtr->fontWeight(500);
        $this->assertEquals(Avtr::URL . 'png?font_weight=500', $avtr->toUrl());
        $avtr->fontWeight(600);
        $this->assertEquals(Avtr::URL . 'png?font_weight=600', $avtr->toUrl());
        $avtr->fontWeight(700);
        $this->assertEquals(Avtr::URL . 'png?font_weight=700', $avtr->toUrl());
        $avtr->fontWeight(800);
        $this->assertEquals(Avtr::URL . 'png?font_weight=800', $avtr->toUrl());
        $avtr->fontWeight(900);
        $this->assertEquals(Avtr::URL . 'png?font_weight=900', $avtr->toUrl());
        $avtr->fontWeight(10000000);
        $this->assertEquals(Avtr::URL . 'png?font_weight=900', $avtr->toUrl());
        $avtr->fontWeight(-10000000);
        $this->assertEquals(Avtr::URL . 'png?font_weight=100', $avtr->toUrl());
    }

    public function testSettingFont()
    {
        $avtr = (new Avtr)->font('open-sans');
        $this->assertEquals(Avtr::URL . 'png?font=open-sans', $avtr->toUrl());
        $avtr->font('source-sans-pro');
        $this->assertEquals(Avtr::URL . 'png?font=source-sans-pro', $avtr->toUrl());
        $avtr->font('roboto');
        $this->assertEquals(Avtr::URL . 'png?font=roboto', $avtr->toUrl());
        $this->expectException(InvalidFontException::class);
        $avtr->font('invalid font');
    }

    public function testAll()
    {
        $avtr = (new Avtr('Jamosaur'))->font('open-sans')
            ->fontWeight(100)
            ->textCase('lower')
            ->theme('flat')
            ->shape('square')
            ->roundedCorners(1)
            ->size(500)
            ->format('jpg')
            ->background(0, 0, 0, 1)
            ->color(255, 69, 0, 1);

        $this->assertEquals(Avtr::URL . 'jpg?initials=Jamosaur&background=rgba%280%2C0%2C0%2C1%29&size=500' .
            '&rounded_corners=1&shape=square&theme=flat&text_case=lower&text_color=rgba%28255%2C69%2C0%2C1%29'.
        '&font_weight=100&font=open-sans', $avtr->toUrl());
    }
}


src/Avtr.php
<?php declare(strict_types=1);

namespace Jamosaur\Avtr;

use Jamosaur\Avtr\Exceptions\InvalidFontException;
use Jamosaur\Avtr\Exceptions\InvalidFormatException;
use Jamosaur\Avtr\Exceptions\InvalidShapeException;
use Jamosaur\Avtr\Exceptions\InvalidTextCaseException;
use Jamosaur\Avtr\Exceptions\InvalidThemeException;

/**
 * Class Avtr
 *
 * @package Jamosaur\Avtr
 * @author  James Wallen-Jones <j.wallen.jones@googlemail.com>
 */
class Avtr
{
    /**
     * URL For the API.
     */
    const URL = 'https://avtr.io/avtr.';

    /**
     * Config parameters.
     *
     * @var array
     */
    protected $config = [
        'initials'        => null,
        'first_name'      => null,
        'last_name'       => null,
        'email'           => null,
        'letter_count'    => null,
        'background'      => null,
        'size'            => null,
        'rounded_corners' => null,
        'shape'           => null,
        'theme'           => null,
        'text_case'       => null,
        'text_color'      => null,
        'font_weight'     => null,
        'font'            => null,
    ];

    /**
     * Default Format that should be returned.
     *
     * @var string
     */
    protected $format = 'png';

    /**
     * Avtr constructor.
     *
     * @param string $value
     */
    public function __construct(string $value = '')
    {
        if (strpos($value, '@')) {
            $this->config['email'] = $value;
        } elseif (str_word_count($value) > 1) {
            $this->config['first_name'] = explode(' ', $value)[0];
            $this->config['last_name']  = explode(' ', $value)[1];
        } else {
            $this->config['initials'] = (($value === '') ? null : $value);
        }
    }

    /**
     * Returns the URL.
     *
     * @return String
     */
    public function toUrl(): String
    {
        $values = count(array_filter($this->config));

        return self::URL . $this->format . (($values > 0) ? '?' . http_build_query($this->config) : null);
    }

    /**
     * Sets the format that should be returned.
     *
     * @param string $format
     *
     * @return Avtr
     * @throws InvalidFormatException
     */
    public function format(string $format): Avtr
    {
        $validFormats = ['png', 'jpg', 'gif'];
        if (!in_array($format, $validFormats)) {
            throw new InvalidFormatException;
        }
        $this->format = $format;

        return $this;
    }

    /**
     * Sets the first name.
     *
     * @param string $firstName
     *
     * @return Avtr
     */
    public function firstName(string $firstName): Avtr
    {
        $this->config['first_name'] = $firstName;

        return $this;
    }

    /**
     * Sets the last name.
     *
     * @param string $lastName
     *
     * @return Avtr
     */
    public function lastName(string $lastName): Avtr
    {
        $this->config['last_name'] = $lastName;

        return $this;
    }

    /**
     * Sets the letter count.
     *
     * @param int $count
     *
     * @return Avtr
     */
    public function letterCount(int $count): Avtr
    {
        if ($count < 1 || $count > 2) {
            $count = 2;
        }
        $this->config['letter_count'] = $count;

        return $this;
    }

    /**
     * Sets the background colour.
     *
     * @param int   $r
     * @param int   $g
     * @param int   $b
     * @param float $a
     *
     * @return Avtr
     */
    public function background(int $r, int $g, int $b, float $a = 1): Avtr
    {
        $this->setColour('background', $r, $g, $b, $a);

        return $this;
    }

    /**
     * Sets the image size.
     *
     * @param int $size
     *
     * @return Avtr
     */
    public function size(int $size): Avtr
    {
        if ($size < 0) {
            $size = 100;
        }

        $this->config['size'] = $size;

        return $this;
    }

    /**
     * Sets whether or not the square image should have rounded corners.
     *
     * @param bool $roundedCorners
     *
     * @return Avtr
     */
    public function roundedCorners(bool $roundedCorners): Avtr
    {
        $this->config['rounded_corners'] = (int)$roundedCorners;

        return $this;
    }

    /**
     * Sets the image shape.
     *
     * @param string $shape
     *
     * @return Avtr
     * @throws InvalidShapeException
     */
    public function shape(string $shape): Avtr
    {
        if (!in_array($shape, ['square', 'circle'])) {
            throw new InvalidShapeException;
        }

        $this->config['shape'] = $shape;

        return $this;
    }

    /**
     * Sets the theme.
     *
     * @param string $theme
     *
     * @return Avtr
     * @throws InvalidThemeException
     */
    public function theme(string $theme): Avtr
    {
        if (!in_array($theme, ['material', 'flat'])) {
            throw new InvalidThemeException;
        }

        $this->config['theme'] = $theme;

        return $this;
    }

    /**
     * Sets the text case.
     *
     * @param string $textCase
     *
     * @return Avtr
     * @throws InvalidTextCaseException
     */
    public function textCase(string $textCase): Avtr
    {
        if (!in_array($textCase, ['lower', 'upper', 'title'])) {
            throw new InvalidTextCaseException;
        }

        $this->config['text_case'] = $textCase;

        return $this;
    }

    /**
     * Sets the text colour.
     *
     * @param int   $r
     * @param int   $g
     * @param int   $b
     * @param float $a
     *
     * @return Avtr
     */
    public function color(int $r, int $g, int $b, float $a = 1): Avtr
    {
        $this->setColour('text_color', $r, $g, $b, $a);

        return $this;
    }

    /**
     * Sets the font weight.
     *
     * @param int $fontWeight
     *
     * @return Avtr
     */
    public function fontWeight(int $fontWeight): Avtr
    {
        $fontWeight                  = min(max($fontWeight, 100), 900);
        $this->config['font_weight'] = $fontWeight;

        return $this;
    }

    /**
     * Sets the font.
     *
     * @param string $font
     *
     * @return Avtr
     * @throws InvalidFontException
     */
    public function font(string $font): Avtr
    {
        $fonts = ['open-sans', 'source-sans-pro', 'roboto'];

        if (!in_array($font, $fonts)) {
            throw new InvalidFontException;
        }

        $this->config['font'] = $font;

        return $this;
    }

    /**
     * Validates and sets the appropriate config value for colours.
     *
     * @param string $where
     * @param int    $r
     * @param int    $g
     * @param int    $b
     * @param float  $a
     */
    private function setColour(string $where, int $r, int $g, int $b, float $a = 1)
    {
        $r = min(max($r, 0), 255);
        $g = min(max($g, 0), 255);
        $b = min(max($b, 0), 255);
        $a = min(max($a, 0), 1);

        $this->config[$where] = 'rgba(' . $r . ',' . $g . ',' . $b . ',' . $a . ')';
    }
}
